### PR TITLE
feat(commenting): price cluster merges at imprecise pins' visual positions

### PR DESCRIPTION
**Pull request by meg-an31**
In order for comment clustering to fire when markers *look* mergeable, this PR prices merges at the pins' visual positions instead of their raw anchors. Imprecise shape pins render tucked ~20 screen px into their shape — a constant that doesn't scale with zoom — so badges formed while such pins were visibly far apart (up to ~2·Tc), and landed off the pins' visual centroid.

`computeClusterTable` gains an optional third argument: a map of the imprecise pins' render offsets. The replay prices an offset pair's merge at the zoom where the markers are visually Tc apart — the largest positive root of `|z·ΔC + Δō| = Tc`, a closed-form quadratic solved once per candidate pair during the build — and carries the exact Tu crossing as the pair's split threshold. Cluster mean offsets dilute as precise members fold in, the `Dmax` fit cap still applies, and pairs whose visuals never reach Tc simply never merge. Nothing new runs while zooming: the table is still precomputed once and played back by the untouched runtime.

The guarantee, pinned by tests: when the map is omitted, empty, or every pair's mean offsets cancel, pricing takes the original code path unchanged — the output table is deep-equal to a pre-change run. The deliberate approximations: a V-shaped pair (offsets larger than Tc, visuals crossing over) stays merged below its first crossing rather than re-splitting on zoom-out; the fit cap measures raw page-space spread; `eps` contraction can still snap a corrected threshold to its window anchor.

Rendering follows the same model: the cluster badge draws at the centroid plus the members' mean inset (the centroid of the pins as rendered), and a coincident stack's badge applies the shared inset of the pins it stands in for.

### Change type

- [x] `improvement`

### Test plan

1. Configure imprecise comments (`shouldBePrecise: () => false`) and place a comment on a shape plus a precise comment nearby on each side.
2. Zoom out: each pair's badge should appear exactly when the two *pins* close to ~22px, not while they're visibly apart; the badge sits centred on where the pins were.
3. Two imprecise comments on one shape still stack (coincident), and the stack badge sits tucked into the shape like the pin it replaces.
4. Precise-only boards behave identically to before.

- [x] Unit tests

### Release notes

- Comment clustering now measures imprecise pins where they actually render, so badges form and split exactly when pins visually meet the clustering thresholds.

### Code changes

| Section   | LOC change |
| --------- | ---------- |
| Core code | +217 / -37 |
| Tests     | +267 / -13 |

### CODE REVIEW

File: packages/commenting/src/clustering/screen-offsets.test.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
+import type { VecLike } from 'tldraw'
+import { describe, expect, it } from 'vitest'
+import { computeClusterTable } from './computeClusterTable'
+import type { LeafInput } from './types'
+
+function leaf(id: string, x: number, y: number): LeafInput {
+	return { id, point: { x, y } }
+}
+
+function mulberry32(seed: number): () => number {
```
Comment (mimecuvalo) [inline packages/commenting/src/clustering/screen-offsets.test.ts:10]: nit, this `mulberry32` is getting duplicated a lot - time for a shared testing fn :)

**Review by mimecuvalo — COMMENTED** [review]
LG! one AI review comment for you, pre-approving

---

Blocking issue: the exact zSplit breaks the runtime's monotonicity invariant

ClusterRuntimeImpl is a prefix cursor: events[0..k) applied. Three places depend on zSplit being non-increasing down the table, and one of them says so in a comment:

- seedFrom — "zSplit is non-increasing down the table, so everything past the first mandatory split is also mandatorily split — this cut can never conflict with an event that should be merged."
- onCamera's split walk: while (k > 0 && zoom >= events[k-1].zSplit) k--
- seedCount's binary search over sqrt(zMerge · zSplit)

Before this PR the invariant was free: zSplit = zMerge · r (plus order-preserving min clamps), and zMerge is non-increasing by construction (lastZ clamp in the replay). An exact per-pair Tu crossing has no such relationship — band width is now (Tu − Tc)/|ΔC|, so a later, lower-zMerge pair with a small |ΔC| and an outward offset gets a much wider band than an earlier one.

I reproduced this against the PR head. Four comments — one imprecise pin with a ±10px outward inset 20 page units from its neighbour, plus two ordinary precise comments ~147 units apart elsewhere — at Tc/Tu/Dmax defaults:

events[0]  zMerge 0.1500  zSplit 0.1800  seedThreshold 0.1643   (precise pair)
events[1]  zMerge 0.1000  zSplit 0.3200  seedThreshold 0.1789   (offset pair)  ← both inverted

Consequences, all observed:

1. Stale badges. Stepping the camera through onCamera, the precise pair stays merged from zoom 0.180 (its own zSplit) all the way to 0.320, because the split walk hits events[1]'s wide band first and stops. A cluster visibly fails to break apart when you zoom past its threshold.
2. seed() returns a wrong cursor. At zoom 0.170 the binary search yields k = 2; a linear scan yields k = 0. seed() runs on every table rebuild (cluster-model.ts), so this is a cold-start path, not an edge case.
3. Structurally, the prefix cursor cannot represent "events[1] merged, events[0] split" at all — so this isn't only a search bug; the exact-split model needs the table to stay totally ordered.

Frequency at production defaults (default eps: 0.7, 40% of comments imprecise, 8–27 comments, 300 seeded random boards): 252 of 300 boards produce at least one inversion, and the same 252 break sqrt(zMerge·zSplit) monotonicity. This is the common case, not a corner.

The existing 151 clustering tests all pass — the invariant simply has no test. The new screen-offsets.test.ts cases are all 2–3 leaf tables, where ordering can't be observed.

Suggested fix — clamp zSplit monotonically in finalize, mirroring the lastZ clamp the replay already applies to zMerge:

let lastSplit = Number.POSITIVE_INFINITY
for (const event of events) {
  const zMerge = Math.min(event.zMerge, zMergeCap)
  if (zMerge < opts.minZoom) break
  let zSplit = event.zSplit !== undefined && event.zSplit > zMerge
    ? Math.min(event.zSplit, opts.maxSplitZoom)
    : zMerge * r
  if (zMerge < opts.maxZoom) zSplit = Math.min(zSplit, opts.maxZoom)
  zSplit = Math.min(zSplit, lastSplit)   // ← keep the table totally ordered
  lastSplit = zSplit
  ...
}

File: packages/commenting/src/clustering/screen-offsets.test.ts (hunk 1 above)
Comment (meg-an31) [reply packages/commenting/src/clustering/screen-offsets.test.ts:10]: this is fixed in the most recent push! weird edge case so ive just gone for the most naïve approach of fixing it, thank u :)
